lib/Ethereum/RPC/Contract/ContractTransaction.pm

```diff
@@ -0,0 +1,121 @@
+package Ethereum::RPC::Contract::ContractTransaction;
+
+use strict;
+use warnings;
+
+our $VERSION = '0.001';
+
+=head1 NAME
+
+   Ethereum::RPC::Contract::ContractTransaction - Centralize contract transactions
+
+=cut
+
+
+use Moo;
+use Ethereum::RPC::Contract::ContractResponse;
+use Ethereum::RPC::Contract::Helper::UnitConversion;
+
+has contract_address => ( is => 'ro' );
+has rpc_client       => ( is => 'ro', lazy => 1 );
+
+sub _build_rpc_client {
+    return Ethereum::RPC::Client->new;
+}
+
+has data             => ( is => 'ro', required => 1 );
+has from             => ( is => 'ro');
+has gas              => ( is => 'ro');
+has gas_price        => ( is => 'ro');
+
+=head2 call_transaction
+
+Call a public functions and variables from a ethereum contract
+
+Return:
+    Ethereum::RPC::Contract::ContractResponse, error message
+
+=cut
+
+sub call_transaction {
+    my $self = shift;
+
+    my $res = $self->rpc_client->eth_call([{
+        to    => $self->contract_address,
+        data  => $self->data,
+    }, "latest"]);
+
+    return (Ethereum::RPC::Contract::ContractResponse->new({ response => $res }), undef) if $res and $res =~ /^0x/;
+
+    return ( undef, $res );
```

Comment:
I'd recommend against this style of error handling.

For calls which have an immediate return, exceptions would be better. Alternatively, since this involves I/O, [Future](https://metacpan.org/pod/Future) would be a good fit.